internal/exec/exec.go

```diff
@@ -0,0 +1,487 @@
+	// 		return err
+	// 	}
+	// }
+
+	if e.job != nil {
+		if err := e.job.UpdateProcThreadAttribute(siEx.ProcThreadAttributeList); err != nil {
+			return err
+		}
+	}
+
+	var zeroSec windows.SecurityAttributes
+	pSec := &windows.SecurityAttributes{Length: uint32(unsafe.Sizeof(zeroSec)), InheritHandle: 1}
+	tSec := &windows.SecurityAttributes{Length: uint32(unsafe.Sizeof(zeroSec)), InheritHandle: 1}
+
+	siEx.Cb = uint32(unsafe.Sizeof(*siEx))
+	if e.execConfig.token != 0 {
+		err = winapi.CreateProcessAsUser(
+			e.execConfig.token,
+			argv0p,
+			argvp,
+			pSec,
+			tSec,
+			inheritHandles,
+			flags,
+			createEnvBlock(addCriticalEnv(dedupEnvCase(true, e.env))),
+			dirp,
+			&siEx.StartupInfo,
+			pi,
+		)
+	} else {
+		err = windows.CreateProcess(
+			argv0p,
+			argvp,
+			pSec,
+			tSec,
+			inheritHandles,
+			flags,
+			createEnvBlock(addCriticalEnv(dedupEnvCase(true, e.env))),
+			dirp,
+			&siEx.StartupInfo,
+			pi,
+		)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to create process: %w", err)
+	}
+	// Don't need the thread handle for anything.
+	defer windows.CloseHandle(windows.Handle(pi.Thread)) //nolint:errcheck
+
+	// Grab an *os.Process to avoid reinventing the wheel here. The stdlib has great logic around waiting, exit code status/cleanup after a
+	// process has been launched.
+	e.process, err = os.FindProcess(int(pi.ProcessId))
+	if err != nil {
+		return fmt.Errorf("failed to find process after starting: %w", err)
+	}
+	return nil
+}
+
+// Run will run the process to completion. This can be accomplished manually by calling Start + Wait afterwards.
+func (e *Exec) Run() error {
+	if err := e.Start(); err != nil {
+		return err
+	}
+	return e.Wait()
+}
+
+// Close will release resources tied to the process (stdio etc.)
+func (e *Exec) Close() error {
+	if e.procState == nil {
+		return errProcNotFinished
+	}
+	winapi.DeleteProcThreadAttributeList(e.attrList)
+	e.closeStdio()
+	return nil
+}
+
+// Pid returns the pid of the running process. If the process isn't running, this will return -1.
+func (e *Exec) Pid() int {
+	if e.process == nil {
+		return -1
+	}
+	return e.process.Pid
+}
+
+// Exited returns if the process has exited.
+func (e *Exec) Exited() bool {
+	if e.procState == nil {
+		return false
+	}
+	return e.procState.Exited()
+}
+
+// ExitCode returns the exit code of the process. If the process hasn't exited, this will return -1.
+func (e *Exec) ExitCode() int {
+	if e.procState == nil {
+		return -1
+	}
+	return e.procState.ExitCode()
+}
+
+// Wait synchronously waits for the process to complete and will close the stdio pipes afterwards. This should only be called once per Exec
+// object.
+func (e *Exec) Wait() error {
+	if e.process == nil {
+		return errProcNotStarted
+	}
+	if e.waitCalled {
+		return errors.New("exec: Wait was already called")
+	}
+	e.waitCalled = true
+	state, err := e.process.Wait()
+	if err != nil {
+		return err
+	}
+	e.procState = state
```

Comment:
Should there be read/write guards around `procState` and `waitCalled`, since there could be race conditions with their access (ie, if two goroutines call `Wait` at the same time)

Reply:
I'd say locking should be left to lower layers if we needed it, and the fact that we call out in the function comment that this should only be called once seems like a fine deterrent. I just checked how os/exec handles this and it seems there's no locking on the top level `Wait` call either, with a similar "check if this bool has been set to guard against someone calling this twice" semantic. 
https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/os/exec/exec.go;l=499-505